crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs

```diff
@@ -77,9 +77,79 @@ pub(crate) fn unspecified_encoding(checker: &mut Checker, call: &ast::ExprCall)
         .filter(|qualified_name| is_violation(call, qualified_name))
         .map(|qualified_name| (qualified_name.to_string(), Mode::from(&qualified_name)))
     else {
+        return check_pathlib(checker, call);
```

Comment:
I think it would be useful to utilize the `is_violation` function for the same. The qualified name would be `&["pathlib", "Path"]` if I'm not mistaken. This will be useful for cases which involves starred and double starred arguments because currently it gets flagged.

```py
from pathlib import Path

with Path("foo.txt").open(*args) as f:
    pass

with Path("foo.txt").open(**kwargs) as f:
    pass
```

This is a rough sketch of what the APIs could look like:

`is_violation` would be the entry point which would take in a `ExprCall` and a new `Segments` (or a better name) enum. This new enum would be either the segments of the current call expression or an attribute value of the `Path` call expression. The `Segments` enum will be created from the call expression along with the help of the semantic model and it would return `None` if it couldn't in which case the rule function would return.

```rs
// TODO: better name?
enum Segments<'a> {
    Regular(&'a [&'a str]),
    Pathlib(&'a str),
}

impl Segments {
    fn try_from_call(call: &ast::ExprCall, semantic: &SemanticModel) -> Self {}
}

fn is_violation(call: &ast::ExprCall, segments: Segments<'_>) -> bool {
    // If we have something like `*args`, which might contain the encoding argument, abort.
    if call.arguments.args.iter().any(Expr::is_starred_expr) {
        return false;
    }
    // If we have something like `**kwargs`, which might contain the encoding argument, abort.
    if call
        .arguments
        .keywords
        .iter()
        .any(|keyword| keyword.arg.is_none())
    {
        return false;
    }

    match segments {
        Segments::Regular(segments) => match segments {
            ["" | "codecs" | "_io", "open"] => {
                if let Some(mode_arg) = call.arguments.find_argument("mode", 1) {
                    if is_binary_mode(mode_arg).unwrap_or(true) {
                        // binary mode or unknown mode is no violation
                        return false;
                    }
                }
                // else mode not specified, defaults to text mode
                call.arguments.find_argument("encoding", 3).is_none()
            }
            ["tempfile", "TemporaryFile" | "NamedTemporaryFile" | "SpooledTemporaryFile"] => {
                let mode_pos = usize::from(segments[1] == "SpooledTemporaryFile");
                if let Some(mode_arg) = call.arguments.find_argument("mode", mode_pos) {
                    if is_binary_mode(mode_arg).unwrap_or(true) {
                        // binary mode or unknown mode is no violation
                        return false;
                    }
                } else {
                    // defaults to binary mode
                    return false;
                }
                call.arguments
                    .find_argument("encoding", mode_pos + 2)
                    .is_none()
            }
            ["io" | "_io", "TextIOWrapper"] => {
                call.arguments.find_argument("encoding", 1).is_none()
            }
            _ => false,
        },
        Segments::Pathlib(attr) => match attr {
            "open" => {
                if let Some(mode_arg) = call.arguments.find_argument("mode", 0) {
                    if is_binary_mode(mode_arg).unwrap_or(true) {
                        // binary mode or unknown mode is no violation
                        return false;
                    }
                }
                // else mode not specified, defaults to text mode
                call.arguments.find_argument("encoding", 2).is_none()
            }
            "read_text" => call.arguments.find_argument("encoding", 0).is_none(),
            "write_text" => call.arguments.find_argument("encoding", 1).is_none(),
            _ => false,
        },
    }
}
```